### ISSUE TITLE
Performance of batch_size > 1

### ISSUE THREAD

**Issue by hgaiser** [fizyr/keras-retinanet#325]
Theoretically `batch_size > 1` should be working, practically however performance appears to degrade. I've looked at the data generator and loss functions but everything appears to be fine.

I'm not sure where the degradation of performance comes from, perhaps a fresh set of eyes can help uncover the issue? My intuition expects the problem to be in the loss function, or a deeper issue in Keras / Tensorflow perhaps.

In extension, this also breaks multi GPU support since that requires `batch_size > 1`.

@awilliamson I think you ran some tests on this right? Do you still have them stored somewhere? Can you share them?

**Comment by hgaiser** [comment]
Now that https://github.com/fizyr/keras-retinanet/pull/339 is merged, we should re-evaluate if this is still broken. https://github.com/fizyr/keras-retinanet/pull/339 might affect these results. Is anyone interested in testing this? I wouldn't test it in combination with multi-gpu however, only single-gpu.

**Comment by sorinpanduru** [comment]
Hi! 
I can fire up training with batch size 16 today. Will post here if i see anything odd.
Thanks!

**Comment by hgaiser** [comment]
Any update on this?

**Comment by sorinpanduru** [comment]
Hi!
Sorry, I've been busy with other things. 
I only started training yesterday with batch size 10 on my own dataset. Looks pretty ok so far, but still need to redo it with batch size 1 so we can compare.

**Comment by jjiunlin** [comment]
@hgaiser here're my training results of resnet50 on coco.
red: batch_size=1, steeps=5000, lr=1e-5
blue: batch_size=2, steeps=5000, lr=1e-5
Also, I'm training batch_size=4, lr=1e-4 with 2*1080ti(--no-evaluation,evaluation running on cpu is super slow).
I think It works. 
![3](https://user-images.githubusercontent.com/33717622/38261023-ade45ae2-379b-11e8-8237-32fadee29f77.PNG)
![4](https://user-images.githubusercontent.com/33717622/38261026-b0fe8eaa-379b-11e8-9478-6e79340a7f70.PNG)

**Comment by hgaiser** [comment]
Ah awesome. @sorinpanduru let me know if you can confirm the same behaviour. If so, I'll close this issue.

**Comment by yhenon-nextdroid** [comment]
> Ah awesome. @sorinpanduru let me know if you can confirm the same behaviour. If so, I'll close this issue.


I think this is quite premature. All this shows is that using batch_size=2 has some stabilising effect on training vs batch_size=1. 

IMO the issue should not be closed until someone can train a network at a decently large batch_size (say ~8) that converges to the (around) same value as batch_size=1, and in fewer iterations.

**Comment by yhenon** [comment]
I think this is quite premature. All this shows is that using batch_size=2 has some stabilising effect on training vs batch_size=1.

IMO the issue should not be closed until someone can train a network at a decently large batch_size (say ~8) that converges to the (around) same value as batch_size=1, and in fewer iterations.

**Comment by jjiunlin** [comment]
@yhenon I agree with you. 
I get OOM issue when training with batch_size=4 on 2 gpus.
![5](https://user-images.githubusercontent.com/33717622/38263108-bb4141c2-37a1-11e8-99dc-f09118476aed.PNG)

**Comment by rodrigo2019** [comment]
I did a training for my custom dataset with differents batch sizes. My dataset are made by small images, so I changed the default min-side-size and max-size-side calling the script in this way:
"python .\keras_retinanet\bin\train.py --backbone=mobilenet128_1.0 --ima
ge-min-side=200 --image-max-side=400 --epochs=100 --batch-size=8 csv data2train.txt classMap.txt
"
It was trained on a 1070ti, which with batch-size=1 took around 2gb of gpu emmory and 80ms per iteration, with batch-size=8 took 2.7gb and 160ms per iteration

The training time was around 15hrs for each training, here is my results:
![image](https://user-images.githubusercontent.com/8016520/38459946-e31e1f5c-3a86-11e8-8260-b9b2aeb7379f.png)
![image](https://user-images.githubusercontent.com/8016520/38459950-ee9d571c-3a86-11e8-9a62-36d7b11ee5e7.png)
![image](https://user-images.githubusercontent.com/8016520/38459957-ffa7b4ee-3a86-11e8-8f3a-e57405c9d8f7.png)


I also can train with large batch sizes if is needed.
I hope this can help.

**Comment by hgaiser** [comment]
Thanks @rodrigo2019 , those seem to be nice results!

**Comment by sorinpanduru** [comment]
<img width="1295" alt="class_loss" src="https://user-images.githubusercontent.com/1698967/38604176-44178016-3d78-11e8-9083-bf43ce6e6d43.png">
<img width="1288" alt="loss" src="https://user-images.githubusercontent.com/1698967/38604177-44382c9e-3d78-11e8-854b-a09ee98bcf05.png">
<img width="1289" alt="regr_loss" src="https://user-images.githubusercontent.com/1698967/38604178-4457f29a-3d78-11e8-8524-2fd82d7cc8f9.png">

Blue = batch size 1
Orange = batch size 10

Training isn't finished on batch size 1 but it seems it's diverging quite a lot from the batch 10 session.

**Comment by hgaiser** [comment]
What am I looking at? I assume the x-axis is the epoch? Why does it seem to go back in time? Why is the orange line further than the blue line, did you start batch_size 1 later than batch size 10? Is this COCO?

**Comment by hgaiser** [comment]
Ah I see. Strange how your results conflict with those of @rodrigo2019 and @jjiunlin . Silly question, but you're sure you're using the latest version of keras-retinanet?

**Comment by sorinpanduru** [comment]
Yep, it's the latest version... I'll just wait for few more epochs. It might converge at some point

**Comment by jjiunlin** [comment]
That's strange.
5000 steps per epoch for all batch_size = 1,2,4. Not original steps = 10000.
lr = 1e-4, batch_size=4 get mAP=0.312 at epoch 58
![0](https://user-images.githubusercontent.com/33717622/38612988-019466d0-3dbb-11e8-8f5a-0c5372ba2026.png)
![1](https://user-images.githubusercontent.com/33717622/38612993-03e94a36-3dbb-11e8-93c9-33bec37c5608.png)
![2](https://user-images.githubusercontent.com/33717622/38612999-05aaadf6-3dbb-11e8-9b53-661b1f4fff9b.png)
![3](https://user-images.githubusercontent.com/33717622/38613003-075a5bf6-3dbb-11e8-85f1-8f54a052d25e.png)
![4](https://user-images.githubusercontent.com/33717622/38613086-534f1c54-3dbb-11e8-93b2-a470edb2fdbd.png)

**Comment by yhenon** [comment]
Ran a test on pascal VOC, training on trainval07, testing on test07. Learning rate was set to `batch_size * 1e-5`, each epoch was `num_images//batch_size`. Test was run for 50 epochs at batch size 1, 4 & 8.

Results show batching works just fine, that is the batched version converges slightly faster, likely due to increased stability of the gradient update.

![loss_pascal](https://user-images.githubusercontent.com/9943242/38882324-d1ea64e4-4237-11e8-9019-dc790b931b67.png)
![map_pascal](https://user-images.githubusercontent.com/9943242/38882328-d4046dec-4237-11e8-8927-52a4c6e69cb2.png)

**Comment by hgaiser** [comment]
Ah that's good news! I was hoping for more of an improvement w.r.t. bs1, but at least it's working.

**Comment by hgaiser** [comment]
How can the time per epoch be smaller for bs8 than for bs1? Did you scale the number of steps with the batch size? Or did you keep the static 10k steps?

Other than that, does this conclude that `batch_size > 1` is working?

**Comment by yhenon** [comment]
>Did you scale the number of steps with the batch size? 
Indeed I did. I ensured that the same number of images were trained on.

>Other than that, does this conclude that batch_size > 1 is working?
Yeah I believe so, and I suggest we close this. 

If someone has issues with batch_size>1, they can reopen this issue.

**Comment by hgaiser** [comment]
> @yhenon I agree with you.
> I get OOM issue when training with batch_size=4 on 2 gpus.

ps. using the same configuration I just got the same error, presumably because the loss optimization is running on gpu0, it cannot handle the data from all batches and GPUs... not sure how to fix this though.

**Comment by mdfwn** [comment]
Found this issue just now after looking for something related to my problem. I tried training for BS > 1 but as soon as I start the training, Keras indicates ETA for 1 epoch at ~18 hours and I can see that it takes multiple seconds for an iteration of 1 image. When I set it to BS=1, the ETA goes immediately down to ~30 minutes and I can see that 7-10 iterations (i.e. images) are being processed per second. 

Another thing I see is that when I have BS > 1, then the classification and regression losses are a lot higher (like 5x bigger) compared to BS=1 at a given iteration.

**Comment by dabasmoti** [comment]
Is there news about how to solve the performance?

**Comment by swathi24prabhu** [comment]
> Ran a test on pascal VOC, training on trainval07, testing on test07. Learning rate was set to `batch_size * 1e-5`, each epoch was `num_images//batch_size`. Test was run for 50 epochs at batch size 1, 4 & 8.
> 
> Results show batching works just fine, that is the batched version converges slightly faster, likely due to increased stability of the gradient update.
> 
> ![loss_pascal](https://user-images.githubusercontent.com/9943242/38882324-d1ea64e4-4237-11e8-9019-dc790b931b67.png)
> ![map_pascal](https://user-images.githubusercontent.com/9943242/38882328-d4046dec-4237-11e8-8927-52a4c6e69cb2.png)

How to plot this graph?
